### PR TITLE
 refactor: Pre-compute some information 

### PR DESCRIPTION
**Pull request by epage**
My goal is to reduce the release logic to eventually allow
- Doing "anything unchanged" checks before the confirmation prompt
- Include dependency updates  in "anything unchanged" when running in
  dry-run mode
- A single confirmation prompt
- Potentially consolidate commits
- Consolidate pushes

This stopped short of the refactorings needed to unlock these changes
because it was getting to big.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -97,11 +115,86 @@ impl<'m> Package<'m> {
...
+        let prev_tag = {
+            let mut template = Template {
+                prev_version: Some(&prev_version.version_string),
+                version: Some(&prev_version.version_string),
+                crate_name: Some(pkg_meta.name.as_str()),
+                ..Default::default()
+            };
+
+            let tag_prefix = config.tag_prefix(is_root);
+            let tag_prefix = template.render(tag_prefix);
```
Comment (epage) [inline src/main.rs]: We no longer include `date` when rendering `tag_prefix`

File: src/main.rs [hunk 2]
```diff
@@ -97,11 +115,86 @@ impl<'m> Package<'m> {
...
+            template.prefix = Some(&tag_prefix);
+            template.render(config.tag_name())
```
Comment (epage) [inline src/main.rs]: We no longer include `date` when rendering `tag_name`

File: src/main.rs [hunk 3]
```diff
@@ -428,37 +492,57 @@ fn release_package(
...
+    } else {
+        let base = pkg.version.as_ref().unwrap_or_else(|| &pkg.prev_version);
+        let mut template = Template {
+            prev_version: Some(&pkg.prev_version.version_string),
+            version: Some(&base.version_string),
+            crate_name: Some(crate_name),
+            ..Default::default()
+        };
+        let tag_prefix = pkg.config.tag_prefix(pkg.is_root);
+        let tag_prefix = template.render(tag_prefix);
```
Comment (epage) [inline src/main.rs]: We no longer include `date` when rendering `tag_prefix`

File: src/main.rs [hunk 4]
```diff
@@ -428,37 +492,57 @@ fn release_package(
...
+        template.prefix = Some(&tag_prefix);
+
+        let tag_name = template.render(pkg.config.tag_name());
```
Comment (epage) [inline src/main.rs]: We no longer include `date` when rendering `tag_name`

File: src/main.rs [hunk 5]
```diff
@@ -428,37 +492,57 @@ fn release_package(
...
+        let base = pkg.version.as_ref().unwrap_or_else(|| &pkg.prev_version);
+        let template = Template {
+            prev_version: Some(&pkg.prev_version.version_string),
+            version: Some(&base.version_string),
+            crate_name: Some(crate_name),
+            date: Some(now.as_str()),
+            next_version: Some(updated_version_string),
+            ..Default::default()
+        };
+        let commit_msg = template.render(pkg.config.post_release_commit_message());
```
Comment (epage) [inline src/main.rs]: We no longer include `prefix` / `tag_name` when rendering `commit_msg`

File: src/main.rs [hunk 6]
```diff
@@ -378,7 +450,14 @@ fn release_package(
...
 
-        let commit_msg = replace_in(pkg.config.pre_release_commit_message(), &replacements);
+        let template = Template {
+            prev_version: Some(&pkg.prev_version.version_string),
+            version: Some(&new_version_string),
+            crate_name: Some(crate_name),
+            date: Some(now.as_str()),
+            ..Default::default()
+        };
+        let commit_msg = template.render(pkg.config.pre_release_commit_message());
```
Comment (epage) [inline src/main.rs]: We no longer include `prefix` / `tag_name` when rendering `commit_msg`

File: src/main.rs [hunk 7]
```diff
@@ -347,9 +412,16 @@ fn release_package(
 
         if !pkg.config.pre_release_replacements().is_empty() {
             // try replacing text in configured files
+            let template = Template {
+                prev_version: Some(&pkg.prev_version.version_string),
+                version: Some(&new_version_string),
+                crate_name: Some(crate_name),
+                date: Some(now.as_str()),
+                ..Default::default()
```
Comment (epage) [inline src/main.rs]: This didn't have `prefix` / `tag_name` before when doing file replacements.

File: src/main.rs [hunk 8]
```diff
@@ -97,11 +115,86 @@ impl<'m> Package<'m> {
...
+            let tag_prefix = template.render(tag_prefix);
+            template.prefix = Some(&tag_prefix);
+            template.render(config.tag_name())
+        };
+
+        let version = {
+            let mut potential_version = prev_version.version.clone();
+            if args
+                .level
+                .bump_level(&mut potential_version, args.metadata.as_ref())?
```
Comment (epage) [inline src/main.rs]: This call is erroring.  I need to look into why.

File: src/main.rs (hunk 5 above)
Comment (epage) [reply src/main.rs]: This is aligned with the documentation
https://github.com/sunng87/cargo-release/blob/master/docs/reference.md#placeholders

File: src/main.rs (hunk 4 above)
Comment (epage) [reply src/main.rs]: Thoughts on if `{{date}}` should be exposed for these renders?

Comment (sunng87) [reply src/main.rs]: In some case `date` is used as part of version string, but for tag I can't see the need for `{{date}}` alone.

**Pull request merged**